File: 卡码/代码随想录/1/11-图论/Go/dfs/LeetCode_0200.go
// 200. 岛屿数量
package main

func numIslands(grid [][]byte) int {
	dir := [][]int{[]int{0, 1, 0, -1}, []int{1, 0, -1, 0}}
	n, m := len(grid), len(grid[0])
	res := 0
	var dfs func(x, y int)
	dfs = func(x, y int) {
		grid[x][y] = '0'
		for i := 0; i < 4; i++ {
			nextX := x + dir[0][i]
			nextY := y + dir[1][i]
			if nextX < 0 || nextX > n-1 || nextY < 0 || nextY > m-1 {
				continue
			}
			if grid[nextX][nextY] == '1' {
				dfs(nextX, nextY)
			}
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == '1' {
				res++
				dfs(i, j)
			}
		}
	}
	return res
}

File: 卡码/代码随想录/1/11-图论/Js/dfs/LeetCode_0200.js
//200. 岛屿数量
const numIslands = function(grid) {
  const dir=[[0,1],[1,0],[0,-1],[-1,0]];
	let res=0;
	const n=grid.length;
	const m=grid[0].length;
	const dfs=(x,y)=>{
		grid[x][y]="0";
		for(let i=0;i<4;i++){
			const nextX=x+dir[i][0];
			const nextY=y+dir[i][1];
			if(nextX<0||nextX>=n||nextY<0||nextY>=m) continue;
			if(grid[nextX][nextY]=="1") dfs(nextX,nextY);
		}
	}
	for(let i=0;i<n;i++){
		for(let j=0;j<m;j++){
			if(grid[i][j]=="1"){
				res++;
				dfs(i,j);
			}
		}
	}
	return res;
};
